File: Assets/Scripts/MainScripts/DCL/Controllers/PhysicsCast/PhysicsCast.cs
using System.Collections.Generic;
using UnityEngine;
using DCL.Interface;
using DCL.Helpers;
using DCL.Models;

namespace DCL
{
    public class PhysicsCast : Singleton<PhysicsCast>
    {
        private LayerMask layerMaskTarget;

        private IRaycastHandler raycastHandler;

        public PhysicsCast()
        {
            layerMaskTarget = 1 << LayerMask.NameToLayer("OnPointerEvent");
            raycastHandler = new RaycastHandler();
        }

        public void Query(RaycastQuery query)
        {
            switch (query.queryType)
            {
                case "HitFirst":
                    HitFirst(query);
                    break;
                case "HitAll":
                    HitAll(query);
                    break;
            }
        }

        private void HitFirst(RaycastQuery query)
        {
            WebInterface.RaycastHitEntity hitEntity;

            RaycastResultInfo raycastInfo = raycastHandler.Raycast(GetUnityRayFromQuery(query), query.ray.distance, ~layerMaskTarget, query.sceneId);
            WebInterface.RayInfo rayInfo = GetRayInfoFromQuery(query);

            if (raycastInfo != null)
            {
                hitEntity = new WebInterface.RaycastHitEntity()
                {
                    didHit = raycastInfo.hitInfo.isValid,
                    hitNormal = raycastInfo.hitInfo.hit.normal,
                    hitPoint = raycastInfo.hitInfo.hit.point,
                    ray = rayInfo,
                    entity = new WebInterface.HitEntityInfo()
                    {
                        entityId = raycastInfo.hitInfo.collider.entityId,
                        meshName = raycastInfo.hitInfo.collider.meshName
                    }
                };
            }
            else
            {
                hitEntity = new WebInterface.RaycastHitEntity()
                {
                    didHit = false,
                    ray = rayInfo
                };
            }

            WebInterface.ReportRaycastHitFirstResult(query.sceneId, query.queryId, query.queryType, hitEntity);
        }

        private void HitAll(RaycastQuery query)
        {
            WebInterface.RaycastHitEntities raycastHitEntities = new WebInterface.RaycastHitEntities();

            RaycastResultInfoList raycastResults = raycastHandler.RaycastAll(GetUnityRayFromQuery(query), query.ray.distance, ~layerMaskTarget, query.sceneId);

            raycastHitEntities.ray = GetRayInfoFromQuery(query);

            if (raycastResults.hitInfo != null && raycastResults.hitInfo.Length > 0)
            {
                int count = raycastResults.hitInfo.Length;
                List<WebInterface.RaycastHitEntity> hitEntityInfoList = new List<WebInterface.RaycastHitEntity>();

                for (int i = 0; i < count; i++)
                {
                    if (raycastResults.hitInfo[i].isValid)
                    {
                        WebInterface.RaycastHitEntity hitEntity = new WebInterface.RaycastHitEntity();
                        hitEntity.didHit = true;
                        hitEntity.ray = raycastHitEntities.ray;
                        hitEntity.hitPoint = raycastResults.hitInfo[i].hit.point;
                        hitEntity.hitNormal = raycastResults.hitInfo[i].hit.normal;
                        hitEntity.entity = new WebInterface.HitEntityInfo();
                        hitEntity.entity.entityId = raycastResults.hitInfo[i].collider.entityId;
                        hitEntity.entity.meshName = raycastResults.hitInfo[i].collider.meshName;
                        hitEntityInfoList.Add(hitEntity);
                    }
                }

                raycastHitEntities.didHit = true;
                raycastHitEntities.hitPoint = raycastResults.hitInfo[0].hit.point;
                raycastHitEntities.hitNormal = raycastResults.hitInfo[0].hit.normal;
                raycastHitEntities.entities = hitEntityInfoList.ToArray();
            }

            WebInterface.ReportRaycastHitAllResult(query.sceneId, query.queryId, query.queryType, raycastHitEntities);
        }

        private void HitFirstAvatar(RaycastQuery query)
        {

        }

        private void HitAllAvatars(RaycastQuery query)
        {

        }

        private UnityEngine.Ray GetUnityRayFromQuery(RaycastQuery query)
        {
            UnityEngine.Ray ray = new UnityEngine.Ray();

            ray.origin = query.ray.unityOrigin;
            ray.direction = query.ray.direction;

            return ray;
        }

        private WebInterface.RayInfo GetRayInfoFromQuery(RaycastQuery query)
        {
            WebInterface.RayInfo rayInfo = new WebInterface.RayInfo();

            rayInfo.direction = query.ray.direction;
            rayInfo.distance = query.ray.distance;
            rayInfo.origin = query.ray.origin;

            return rayInfo;
        }
    }
}

File: kernel/packages/decentraland-ecs/src/decentraland/PhysicsCast.ts
import { ReadOnlyVector3, Vector3, Matrix } from './math'
import { RaycastResponse } from './Events'
import { uuid, log } from '../ecs/helpers'

/** @internal */
import { DecentralandInterface } from './Types'
import { Camera } from './Camera'

/**
 * @public
 */
export type QueryType = 'HitFirst' | 'HitAll' | 'HitFirstAvatar' | 'HitAllAvatars'

/**
 * @internal
 */
enum QueryPrefix {
  HitFirst = 'rqhf',
  HitAll = 'rqha'
}

/**
 * @internal
 */
export interface RaycastQuery {
  queryId: string
  queryType: QueryType
  ray: Ray
}

/**
 * @public
 */
export interface RaycastHit {
  didHit: boolean
  ray: Ray
  hitPoint: ReadOnlyVector3
  hitNormal: ReadOnlyVector3
}

/**
 * @public
 */
export interface Ray {
  origin: ReadOnlyVector3
  direction: ReadOnlyVector3
  distance: number
}

/**
 * @public
 */
export interface HitEntityInfo {
  isValid: boolean
  entityId: string
  meshName: string
}

/**
 * @public
 */
export interface RaycastHitEntity extends RaycastHit {
  entity: HitEntityInfo
}

/**
 * @public
 */
export interface RaycastHitEntities extends RaycastHit {
  entities: RaycastHitEntity[]
}

/**
 * @public
 */
export interface BasicAvatarInfo {
  userId: string
  name: string
}

/**
 * @public
 */
export interface RaycastHitAvatar extends RaycastHit {
  avatar: BasicAvatarInfo
}

/**
 * @public
 */
export interface RaycastHitAvatars extends RaycastHit {
  avatars: BasicAvatarInfo[]
}

/**
 * @public
 */
export interface IPhysicsCast {
  hitFirst(ray: Ray, hitCallback: (event: RaycastHitEntity) => void, id?: number): void
  hitAll(ray: Ray, hitCallback: (event: RaycastHitEntities) => void, id?: number): void
  /** @internal */
  hitFirstAvatar(ray: Ray, hitCallback: (event: RaycastHitAvatar) => void): void
  /** @internal */
  hitAllAvatars(ray: Ray, hitCallback: (event: RaycastHitAvatars) => void): void
}

/** @internal */
declare let dcl: DecentralandInterface | void

/**
 * @public
 */
export class PhysicsCast implements IPhysicsCast {
  private static _instance: PhysicsCast
  private queries: Record<string, (event: RaycastHit) => void> = {}

  private constructor() {}

  public static get instance(): PhysicsCast {
    PhysicsCast.ensureInstance()
    return PhysicsCast._instance
  }

  static ensureInstance(): any {
    if (!PhysicsCast._instance) {
      PhysicsCast._instance = new PhysicsCast()
    }
  }

  public getRayFromCamera(distance: number) {
    let rotation = Camera.instance.rotation
    let rotationMat: Matrix = Matrix.Identity()
    rotation.toRotationMatrix(rotationMat)
    let direction = Vector3.TransformCoordinates(Vector3.Forward(), rotationMat)

    const ray: Ray = {
      origin: Camera.instance.position,
      direction: direction,
      distance: distance
    }

    return ray
  }

  public getRayFromPositions(from: Vector3, to: Vector3) {
    const direction = to.subtract(from)
    const length = direction.length()

    const ray: Ray = {
      origin: from,
      direction: direction.normalize(),
      distance: length
    }

    return ray
  }

  public hitFirst(ray: Ray, hitCallback: (event: RaycastHitEntity) => void, id?: number) {
    const queryId = typeof id === 'number' ? QueryPrefix.HitFirst + id : uuid()

    this.queries[queryId] = hitCallback as (event: RaycastHit) => void

    dcl && dcl.query('raycast', { queryId, queryType: 'HitFirst', ray })
  }

  public hitAll(ray: Ray, hitCallback: (event: RaycastHitEntities) => void, id?: number) {
    const queryId = typeof id === 'number' ? QueryPrefix.HitAll + id : uuid()

    this.queries[queryId] = hitCallback as (event: RaycastHit) => void

    dcl && dcl.query('raycast', { queryId, queryType: 'HitAll', ray })
  }

  public hitFirstAvatar(ray: Ray, hitCallback: (event: RaycastHitAvatar) => void) {
    log('not implemented yet')
  }

  public hitAllAvatars(ray: Ray, hitCallback: (event: RaycastHitAvatars) => void) {
    log('not implemented yet')
  }

  public handleRaycastHitFirstResponse(response: RaycastResponse<RaycastHitEntity>) {
    this.queries[response.payload.queryId](response.payload.payload)
    delete this.queries[response.payload.queryId]
  }

  public handleRaycastHitAllResponse(response: RaycastResponse<RaycastHitEntities>) {
    this.queries[response.payload.queryId](response.payload.payload)
    delete this.queries[response.payload.queryId]
  }
}
